### ISSUE TITLE
StyleGAN: non-4x4-const model

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#21]
In Imagenet and Danbooru and Flickr100m experiments, StyleGAN's poor results on LSUN CATS & CARS (from Karras et al) have been replicated. Even given a ridiculous amount of TPU time, StyleGAN is just unable to model multiple objects well; and the FIDs are terrible - eg >100 for ImageNet (whereas BigGAN achieves <100 within a few thousand iterations). Why is StyleGAN so terrible at complex scenes with multiple objects when it achieves such shockingly photorealistic results on datasets like FFHQ which have a single centered object?

Aside from the 8x512 FC layers and the AdaIn upscaling layers used in the Laplacian pyramid, there's a major difference between StyleGAN and BigGAN/most other GANs: the upscaling CNN starts with a single learned constant 4x4 base input, which is then upscaled, with the latent embedding _w_ injected each layer along with noise.

The additive 'upscaling' of a learned constant may explain why it does central objects and anime so well: it can learn a single seed object, and it can 'sculpt' the object as it upscales. StyleGAN appears to naturally factorize image generation, eg when we visualize each resolution as an RGB image, we can see that for anime-style drawings it has separate layers for lines and for colors, which then combine to give an image with lines & filled regions of colors.

However, in a case of bias/variance and perhaps 'the bitter lesson', the disconnect between the learned constant input & the _w_ embedding weakens it drastically compared to BigGAN (which instead learns global structure well, but struggles with texture and finer details).

Experiments: aydao experimented with making the 4x4 constant input depend on the _w_ latent vector (adding in 1 or 2 layers as an adapter), making it like other GANs. On the very simple Flowers dataset, this worked OK, but then when tested on the far more complex Danbooru, this destabilized & failed to train; however, inserting the adapter layers into a pre-trained Danbooru StyleGAN model worked, recovering within a few thousand iterations (but without substantial quality or multi-object modeling improvements).

l4rz built on this for his [topless StyleGAN](https://github.com/l4rz/practical-aspects-of-stylegan2-training) (config-f) and instead of training the entire model, he froze most of it (a standard transfer learning trick), and focused training on the adapter layers, gradually unfreezing layers ([video](http://l4rz.net/training-36590-36918-4x4-noconst-initalization.mp4)). He reports that this gives quite different results and has a clear increase in variety. After a few days the StyleGAN had recovered. aydao may try Danbooru again.

---

This is interesting for giving insight into why StyleGAN works and what price it pays for its spectacular performance, and also into a way around it: train StyleGAN quickly to the limits of a single const input, then do model surgery to add in adapter layers, fade those in, and then unlock and train the rest of it. It would be interesting to see to what extent a full non-const StyleGAN can compete with BigGAN (lacking self-attention or specnorms but having a very deep embedding & AdaIn layers).

**Comment by gwern** [comment]
Further update: l4rz reports no learning on his topless model despite increasing LR. He is skeptical that no-const is doing anything useful. Bit of a puzzle here: it may be unstable from scratch, but then when converged, it's frozen in place---does it need much higher LRs/# of iterations to adapt during transfer learning, or is it just too late?

**Comment by gwern** [comment]
Another datapoint which bears on the question of what StyleGAN does, exactly, that makes it so good at face-like datasets but fail to scale: Aydao has experimented with a BigGAN with _w_ (#26 ), minibatch stddev, and with the style modulation/transfer layers rather than pure convolutions. The result is... it looks just like a regular BigGAN _w_, aside from looking more StyleGAN-blob-like early on.

One hypothesis here is that the style transfer layers don't really do anything useful, because they are redundant with Batchnorm in terms of normalizing layers/inputs/outputs. Karras has always made a point of how StyleGAN doesn't *need* BN (implying it would get even better if one had the bad taste to add in BN), but it looks like it's simply because StyleGAN uses BN in a different form, as it were, and so when you drop in the signature StyleGAN layers to an architecture which already has BN, it's redundant.

So, since the style transfer layers add nothing, and _w_ layers empower BigGAN to handle multimodal stuff further, that further suggests that the const input + Laplacian Pyramid may be part of why StyleGAN hits that particular tradeoff.
